Bulk Publishing: Replace calls to `datapointsStore.getDatapoints(agencyId)` w/ the updated `get_reports_by_id` endpoint

mxosman opened the pull request:
## Description of the change

Calling the `/datapoints` endpoint is too expensive for what we need to power the review pages. The updated `get_reports_by_id` should return a list of datapoints (instead of metrics) for each report provided to it. This way, we only get the relevant datapoints we need for the reports that have been selected by the user and not the entire set of datapoints.

## Related issues

Closes #529

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

mxosman commented on publisher/src/stores/ReportStore.ts:
```diff
@@ -180,9 +180,7 @@ class ReportStore {
 
-      reports.forEach((report) => {
-        this.storeMetricDetails(report.id, report.metrics);
-      });
```
We don't need to store anything in the store regarding these reports - they should already exist within the store.

lilidworkin replied:
Totally makes sense!

lilidworkin reviewed (approved):
Looks totally correct to me! Weird about the bug you're running into -- not sure if it's a FE or BE thing -- after looking about both PRs I don't see an obvious candidate. Spend a bit more time on it and then we can revisit tomorrow.

mxosman commented:
> Looks totally correct to me! Weird about the bug you're running into -- not sure if it's a FE or BE thing -- after looking about both PRs I don't see an obvious candidate. Spend a bit more time on it and then we can revisit tomorrow.

I _think_ I know what's going on... So, it looks like we're using `enabledMetrics` as a source of truth for all of the metrics via:

```
  const enabledMetrics = reportStore.agencyMetrics.filter(
    (metric) => metric.enabled
  );
```

The trouble with this is -- say we have an Annual report with 1 metric enabled and a month report with 2 metrics enabled. When you select a single report (annual or monthly) -- the left hand side list of metrics shows all of the metrics, and oddly renders the page with blank metrics.

Example Annual Report with Funding as its only enabled metric:
<img width="1728" alt="Screenshot 2023-03-30 at 2 13 50 PM" src="https://user-images.githubusercontent.com/59492998/228940284-8d0e9488-efd0-46fe-9ea6-e68b20994b67.png">
The other two metrics are enabled for monthly reports.

I'm going to first work on the backend of this and incorporate your suggestions and then come back and think of a better way to get a list of `enabledMetrics` relevant to the set of reports a user has selected.

lilidworkin commented:
@mxosman interesting, I think I understand! Actually, is this why Ivan was doing this:

```
reports.forEach((report) => {
    this.storeMetricDetails(report.id, report.metrics);
});
```

Now that I think about it, we _do_ actually need to do this, because while the store already knows about all reports, it only knows the report overview details -- it doesn't know about the report metrics. Right?

And in fact, it's more complicated than that, because we just changed the backend so that `report.metrics` is no longer populated, only `report.datapoints`. But it looks like we might actually need both to be populated. I think if you 1) change the backend so that `report.metrics` and `report.datapoints` are both populated, and 2) put back this `storeMetricDetails` logic, then it should work -- can you test that?

mxosman commented:
> @mxosman interesting, I think I understand! Actually, is this why Ivan was doing this:
> 
> ```
> reports.forEach((report) => {
>     this.storeMetricDetails(report.id, report.metrics);
> });
> ```
> 
> Now that I think about it, we _do_ actually need to do this, because while the store already knows about all reports, it only knows the report overview details -- it doesn't know about the report metrics. Right?
> 
> And in fact, it's more complicated than that, because we just changed the backend so that `report.metrics` is no longer populated, only `report.datapoints`. But it looks like we might actually need both to be populated. I think if you 1) change the backend so that `report.metrics` and `report.datapoints` are both populated, and 2) put back this `storeMetricDetails` logic, then it should work -- can you test that?

Yeah, I think you're right... bummer! I actually wrote and removed that snippet above. The interesting thing is... we are getting a list of the metrics via the `/metrics` endpoint and storing them in the ReportStore with this function `initializeReportSettings`. So we might be OK not needing to update the endpoint to send the list of metrics. One thing I might try is to filter the list of enabled reports further to exclude metrics w/ frequencies that don't match the list of reports the user selected? Might not be pretty, but will at least solve the left hand-side list of metrics showing all enabled metrics regardless of whether or not you selected only monthly reports/only selected annual reports. If you select a monthly + annual report to publish, then it should show you all the enabled metrics. Does that make sense?

Let me try a few things first before readjusting the backend to also send metrics!

lilidworkin commented:
Hmm @mxosman that sort of makes sense but I also feel like I'm getting confused here. In order to determine what metrics show up on the left side, don't we want to just check what metrics appear in the list of datapoints that the BE has sent over? For instance, if `get_reports_by_ids` returns something that looks like this:

```
[
     {
          "report 1",
          "datapoints": [ {"funding": 100} ]
     },
     {
          "report 2",
          "datapoints": [ {"arrests": 5} ]
     },
     {},   
]
```

Then we should show funding and arrests in the left side bar. So I guess I'm confused about why we need to reference anything else that already exists in the store. But I think I'm probably missing something.

mxosman commented:
> Hmm @mxosman that sort of makes sense but I also feel like I'm getting confused here. In order to determine what metrics show up on the left side, don't we want to just check what metrics appear in the list of datapoints that the BE has sent over? For instance, if `get_reports_by_ids` returns something that looks like this:
> 
> ```
> [
>      {
>           "report 1",
>           "datapoints": [ {"funding": 100} ]
>      },
>      {
>           "report 2",
>           "datapoints": [ {"arrests": 5} ]
>      },
>      {},   
> ]
> ```
> 
> Then we should show funding and arrests in the left side bar. So I guess I'm confused about why we need to reference anything else that already exists in the store. But I think I'm probably missing something.

Sorry, I didn't do a great job of explaining -

```
  const enabledMetrics = reportStore.agencyMetrics.filter(
    (metric) => metric.enabled
  );
```

This snippet is currently what's being used to render that metrics list -- it's taking the entire list of metrics and filtering out the disabled ones. We need to replace this with something that will just show the list of enabled metrics for the reports that were selected because this will not help us unless we do further filtering that's a little wonky.

I just came to the same conclusion as you! I think it makes sense to loop through all the datapoints and create a set of the metrics provided by the datapoints.

mxosman commented:
This might be tricky because we're receiving the full list of datapoints including disabled metrics. I guess we can have the FE go through and give us a list of enabledMetricKeys and use that to filter the datapoints. But, now I'm wondering if it's just easier to update the backend to send a list of metrics -- really all we need from a list of metrics (now that we have the datapoints) are a list of the enabled ones - we get everything else from the datapoints.

mxosman commented:
> This might be tricky because we're receiving the full list of datapoints including disabled metrics. I guess we can have the FE go through and give us a list of enabledMetricKeys and use that to filter the datapoints. But, now I'm wondering if it's just easier to update the backend to send a list of metrics -- really all we need from a list of metrics (now that we have the datapoints) are a list of the enabled ones - we get everything else from the datapoints.

Annual + Monthly Report:
<img width="1093" alt="Screenshot 2023-03-31 at 10 41 57 AM" src="https://user-images.githubusercontent.com/59492998/229167192-95a62833-ffaf-4ac7-b214-566658efbe58.png">

Datapoints for both reports:
<img width="600" alt="Screenshot 2023-03-31 at 10 42 02 AM" src="https://user-images.githubusercontent.com/59492998/229167255-c29d2272-b66f-484f-b3df-75b627361af3.png">

lilidworkin commented on publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:
```diff
@@ -202,11 +205,19 @@ export const ReviewMetrics: React.FC<ReviewMetricsProps> = ({
+
+      {hasNoDatapoints ? (
+        <NoDatapointsMessage>
+          There are no metric data to review. Please go back and add metric data
```
nit, how about "There is no data to review. Please go back and enter data in order to Review & Publish."

mxosman replied:
I love that so much! Thank you!

lilidworkin commented on publisher/src/components/Reports/BulkActionReview.tsx:
```diff
@@ -99,26 +117,44 @@ const BulkActionReview = () => {
+    return keys;
+  }, [] as string[]);
+
+  const metricsToDisplay = Object.entries(datapoints).reduce(
```
Nice! So we're basically taking an intersection of enabled metrics and the metrics that appear in the list of datapoints. Seems perfect!

mxosman replied:
Exactly!

lilidworkin reviewed (approved):
Your solution looks great!
